### PR TITLE
Initializer Improvements

### PR DESCRIPTION
**Pull request by illiakovalenko**
## Description / Motivation

* enhancements to transform customizable/extendable "COPY_ONLY_REGEX" useable with 'transform' filter, filter ащк files to only be copied (binaries, .jar, vue's index.html) vs files to be pruned/not rendered (empty language .yml)
* Fix lint warnings
* Remove "src\common\steps\index.ts"
* Rename "src\common\steps" to "src\common\processes"
* Create a "src\common\index.ts" that exports anything which is expected to be used by initializer implementations, update initializers to use these (instead of full path for import):
* ^ add JSDoc comments to the above 'public' functions
*  Decided to use `types` as they have shorter syntax and look better, and see the difference: https://stackoverflow.com/a/37233777
* remove `isJssApp` as we allow customer to decided how they want to use post-initializers

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/create-sitecore-jss/src/common/Initializer.ts [hunk 1]
```diff
@@ -1,11 +1,22 @@
...
   initializers?: string[];
   nextSteps?: string[];
-}
-export interface Initializer {
+};
+
+/**
+ * Initializer base type
+ */
+export type Initializer = {
```
Comment (ambrauer) [inline packages/create-sitecore-jss/src/common/Initializer.ts:12]: For class interfaces, I think we should stick to `interface` (common in OO languages,[ TS docs](https://www.typescriptlang.org/docs/handbook/2/classes.html#implements-clauses) seem to prefer this too).

File: packages/create-sitecore-jss/src/common/processes/transform.ts [hunk 2]
```diff
@@ -147,11 +147,42 @@ export const diffAndWriteFiles = async ({
   }
 };
 
+type TransformOptions = {
+  /**
+   * Determines whether a file should be copied
```
Comment (ambrauer) [inline packages/create-sitecore-jss/src/common/processes/transform.ts:152]: Name of this option might be misleading. It's more a "is file for copy **only**". Maybe we just add that to the doc.
```suggestion
   * Determines whether a file should be copied only (not rendered through ejs)
```

File: packages/create-sitecore-jss/src/common/processes/transform.ts [hunk 3]
```diff
@@ -147,11 +147,42 @@ export const diffAndWriteFiles = async ({
...
+  isFileForCopy?: (file: string, fileForCopyRegExp: RegExp) => boolean;
+  /**
+   * Determines whether a file should be skiped (not copied/rendered).
+   * @param {string} file path to a file
+   */
+  isFileForSkip?: (file: string) => boolean;
+  /**
+   * Custom RegExp to determine which files should be copied without render
+   */
+  fileForCopyRegExp?: RegExp;
```
Comment (ambrauer) [inline packages/create-sitecore-jss/src/common/processes/transform.ts:165]: Is it confusing to offer both `isFileForCopy` and `fileForCopyRegExp` options? If they needed to override the RegExp, they could technically do that with `isFileForCopy`.

File: packages/create-sitecore-jss/src/common/utils/helpers.ts [hunk 4]
```diff
@@ -54,6 +25,11 @@ export const getPascalCaseName = (name: string): string => {
   return name;
 };
 
+/**
+ * Provides `package.json` data
+ * @param {string} [pkgPath] path to `package.json`
```
Comment (ambrauer) [inline packages/create-sitecore-jss/src/common/utils/helpers.ts:30]:
```suggestion
 * @param {string} [pkgPath] path to `package.json`. Default is './package.json'.
```

File: packages/create-sitecore-jss/src/common/utils/helpers.ts [hunk 5]
```diff
@@ -65,10 +41,15 @@ export const openPackageJson = (pkgPath?: string) => {
   }
 };
 
-export const writePackageJson = (object: unknown, pkgPath?: string) => {
+/**
+ * Creates `package.json` file and inserts provided data
+ * @param {Object} data data to be written into package.json
+ * @param {string} [pkgPath] a path to a file
```
Comment (ambrauer) [inline packages/create-sitecore-jss/src/common/utils/helpers.ts:47]:
```suggestion
 * @param {string} [pkgPath] a path to a file. Default is './package.json'.
```

**Review by ambrauer — CHANGES_REQUESTED** [review]
This looks great, just a few comments.

File: packages/create-sitecore-jss/src/common/processes/transform.ts (hunk 3 above)
Comment (illiakovalenko) [reply packages/create-sitecore-jss/src/common/processes/transform.ts:165]: I thought about it if you want just to override default RegExp, without `isFileForCopy` usage. And it's possible that they will want to use our RegExp and add some additional logic, in this case I provided this regexp as the second parameter to the function

File: packages/create-sitecore-jss/src/common/processes/transform.ts (hunk 2 above)
Comment (illiakovalenko) [reply packages/create-sitecore-jss/src/common/processes/transform.ts]: Done

File: packages/create-sitecore-jss/src/common/utils/helpers.ts (hunk 4 above)
Comment (illiakovalenko) [reply packages/create-sitecore-jss/src/common/utils/helpers.ts]: Done

File: packages/create-sitecore-jss/src/common/utils/helpers.ts (hunk 5 above)
Comment (illiakovalenko) [reply packages/create-sitecore-jss/src/common/utils/helpers.ts]: Done

File: packages/create-sitecore-jss/src/common/processes/transform.ts (hunk 3 above)
Comment (illiakovalenko) [reply packages/create-sitecore-jss/src/common/processes/transform.ts:165]: It's problematically to extend our internal regexp if you need it, it was an original issue

Comment (ambrauer) [reply packages/create-sitecore-jss/src/common/processes/transform.ts:167]: Agree, I think it's good as-is, especially with the additional doc comment you added to `isFileForCopy`.

**Review by ambrauer — APPROVED** [review]

**Pull request merged**